lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
+  tryConvertingReturnValue(const SortedVector<RegisterType> &Used,
+                           const RegisterArray<RegisterCount> &Allowed,
+                           const model::Register::Values PointerToCopyLocation,
+                           const model::Binary &Binary,
+                           TypeVector &TypeDependencies) {
+    if (Used.size() == 0) {
+      // Short-circuit the functions that don't return anything (`void`).
+      return ensurePrimitiveTypeIsAvailable(model::PrimitiveTypeKind::Void,
+                                            0,
+                                            Binary,
+                                            TypeDependencies);
+    }
+
+    if (Allowed.empty()) {
+      // A return value is expected on an ABI that doesn't dedicate any
+      // registers to be used for those.
+      return std::nullopt;
     }
 
-    if (UsedRegisters.size() == 1) {
-      if (UsedRegisters.begin()->Location == PointerToCopyLocation) {
-        if constexpr (DryRun)
-          return model::QualifiedType{};
-        else
-          return getTypeOrDefault(UsedRegisters.begin()->Type,
-                                  PointerToCopyLocation,
-                                  TheBinary);
+    // Special case where the only used register is the pointer-to-copy location
+    if (Used.size() == 1 && Used.begin()->Location == PointerToCopyLocation) {
+      if (Used.begin()->Type.isPointer()) {
+        // If the type is already a pointer, preserve it.
+        return Used.begin()->Type;
       } else {
-        if constexpr (RegisterCount == 0)
-          return std::nullopt;
-        if (AllowedRegisters.front() == UsedRegisters.begin()->Location) {
-          if constexpr (DryRun)
-            return model::QualifiedType{};
-          else
-            return getTypeOrDefault(UsedRegisters.begin()->Type,
-                                    UsedRegisters.begin()->Location,
-                                    TheBinary);
-        } else {
-          return std::nullopt;
-        }
+        // Otherwise return a `void *` pointer as the return value type.
+        static constexpr auto VoidKind = model::PrimitiveTypeKind::Void;
+        auto Result = ensurePrimitiveTypeIsAvailable(VoidKind,
+                                                     0,
+                                                     Binary,
+                                                     TypeDependencies);
+        using model::Qualifier;
+        Result.Qualifiers.emplace_back(Qualifier::createPointer(Architecture));
+        return Result;
+      }
+    }
+
+    // TODO: there are some cases where preserving the total structure of
+    // the return value can be benefitial, this needs more work.
+    // For now, just determine the total size of the return value and if it
+    // fits into a primitive value - emit it, otherwise show those values as
+    // a struct (/note: positional ABIs need a whole lot of extra love in this
+    // regard).
```

Comment:
Can you make a specific example? I'm not sure I understand.

```suggestion
    // Note: there are some cases where preserving the total structure of
    // the return value can be beneficial, this needs more work.
    // For now, just determine the total size of the return value and, if it
    // fits into a primitive value - emit it, otherwise show those values as
    // a struct.
    //
    // Note: positional ABIs need a whole lot of extra love in this regard.
```

Reply:
This part is mostly reworked already (it's related to some of the problems on the revng-c side we investigated with Pietro). This approach was too naïve and in essence was only needed because the Layout only even had a single return type (technically, the same can be said about CABI-FT, but it's a lot easier to emulate it using a struct there).

So, don't worry about this for now, next iteration should be a lot easier to understand. I'll close this thread, since it's most likely come up again in the review of the updated return value handling.